### DISCUSSION TITLE
[Question][Hw5] Definition of concrete domains

### DISCUSSION THREAD

**Discussion by p51lee** [prosyslab-classroom/cs524-program-analysis#331] [Q&A]
Hello, I have a question about the definition of concrete domains in our homework.

<img width="1409" alt="image" src="https://user-images.githubusercontent.com/68288688/203566898-35c13f2d-d74f-4f3d-b9eb-aedc5d430964.png">

These concrete domains are context-insensitive and I think it forces to use strong update instead of weak update in some cases.

For example, abstract *filter* function

<img width="911" alt="image" src="https://user-images.githubusercontent.com/68288688/203566085-b3c76e1f-2895-46d6-af14-e74a977ce98b.png">

is always a strong update (simple overwriting) because the concrete domains are context-insensitive. Is it okay to use this definition in our homework implementation?

**Comment by KihongHeo** [comment]
Hi,
First of all, we don't consider the context sensitivity of concrete semantics. Concrete semantics is trivially context-sensitive because every function call is distinguished. Context-sensitivity is a part of abstract semantics. In the above definition, my point is to omit complex definitions such as instances, environments, and continuations because they all will be abstracted in a context insensitive way.

In the filter function, you should think about the soundness by yourself. The filter function describes the behavior of how the filter works, but does not tell you **when** this is applied. This is what you have to think.

**Comment by p51lee** [comment]
Thank you for your explanation 😊
